### PR TITLE
python3-Cython: update to 3.0.1.

### PR DESCRIPTION
**Pull request by ahesford**
This is a big update that reflects several years of extensive work, and needs some testing to be sure it doesn't break any Void packages.

#### Testing the changes
- I tested the changes in this PR: **in progress**

The following packages need to be fixed to build with new Cython:
- [ ] gbinder-python
- [ ] jack_mixer
- [ ] libimobiledevice
- [ ] memray
- [ ] pyliblo
- [x] python3-cypari2
- [x] python3-cysignals
- [ ] python3-efl
- [x] python3-fpylll
- [ ] python3-grpcio
- [ ] python3-h5py
- [ ] python3-jq
- [ ] python3-logbook
- [x] python3-lxml
- [ ] python3-mpi4py
- [ ] python3-pandas-msgpack
- [ ] python3-peewee
- [x] python3-pplpy
- [ ] python3-pyFFTW
- [ ] python3-pygame
- [x] python3-scikit-learn
- [x] python3-scipy
- [ ] python3-shapely
- [ ] python3-tables
- [ ] python3-uvloop
- [x] python3-yaml
- [ ] sagemath
- [ ] urh
- [ ] vapoursynth
- [ ] vidcutter

### CODE REVIEW

**Comment by tornaria** [comment]
Related: https://github.com/sagemath/sage/issues/29863

I'll have a look into (at least) my pkgs: cypari2, cysignals, fpylll, pplpy, sagemath. I can also have a look into scipy as it's needed for sagemath.

Question: could we have cython2 and cython3 coexist for a while?

**Comment by ahesford** [comment]
The two versions of Cython can't coexist without a lot of pain because they will conflict in the Python package namespace. We can just hold this update until everything is resolved.

**Comment by icp1994** [comment]
pyyaml realeased v6.0.1 which [blocks](https://github.com/yaml/pyyaml/commit/ae08bdc82b4ddfcd2b93c8aedcd1963766c3307d) building with Cython 3.0.0

**Comment by ahesford** [comment]
Yeah, I'll bump this and patch around it if necessary.

**Comment by tornaria** [comment]
scipy needs patching to build with pybind 2.11 (or maybe `--skip-dependency-check` since all the patches are about relaxing dependencies (cython, numpy, and now pybind11).

Note that in their main branch they have relaxed the upper bound for cython and pybind11 (they still fix numpy to a particular version). See: https://github.com/scipy/scipy/blob/main/pyproject.toml

File: srcpkgs/python3-fpylll/patches/cython3-legacy.patch [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
++++ b/setup.py
+@@ -123,7 +123,12 @@ class build_ext(_build_ext, object):
+             self.extensions,
+             include_path=["src"],
+             build_dir=self.cythonize_dir,
+-            compiler_directives={"binding": True, "embedsignature": True, "language_level": 2},
++            compiler_directives={
++                "binding": True,
++                "embedsignature": True,
++                "language_level": 2
```
Comment (tornaria) [inline srcpkgs/python3-fpylll/patches/cython3-legacy.patch:19]:
```suggestion
+                "language_level": 2,
```
My bad, that's what I get for hand-editing the patch after checking... This should fix it.

**Comment by ahesford** [comment]
@tornaria I decided it wasn't worth trying to patch around all of these packages and added a new `python3-Cython0.29` which conflicts with `python3-Cython` and offers the "legacy" Cython 0.29.36. The legacy package can be a build-time dependency for the ~30 packages that have yet to support Cython 3. Only two packages have run-time dependencies on Cython: `urh` and `sagemath`. I have moved both the build-time and run-time dependencies of `urh` to `python3-Cython0.29` for now. How should I handle `sagemath`? We can do the same if you think that is the best course of action.

**Comment by tornaria** [comment]
> @tornaria I decided it wasn't worth trying to patch around all of these packages and added a new `python3-Cython0.29` which conflicts with `python3-Cython` and offers the "legacy" Cython 0.29.36. The legacy package can be a build-time dependency for the ~30 packages that have yet to support Cython 3. Only two packages have run-time dependencies on Cython: `urh` and `sagemath`. I have moved both the build-time and run-time dependencies of `urh` to `python3-Cython0.29` for now. How should I handle `sagemath`? We can do the same if you think that is the best course of action.

I think we have a way forward with sagemath (https://github.com/sagemath/sage/pull/36109 is almost done with review and https://github.com/sagemath/sage/pull/36110 is small). I think it might be possible to have sagemath running on cython 3 within the week.

**Comment by ahesford** [comment]
With `python3-pplpy` built with `python3-Cython0.29` (which is its current state in the repository), all `sagemath` tests in the `--long` suite pass (with the exception of some new warnings emitted by Pythran), so this is good to go; merged as cb4c843f1feb13347fb472aa71d18f450a0a5413.

**Comment by tornaria** [comment]
> With `python3-pplpy` built with `python3-Cython0.29` (which is its current state in the repository), all `sagemath` tests in the `--long` suite pass (with the exception of some new warnings emitted by Pythran), so this is good to go; merged as [cb4c843](https://github.com/void-linux/void-packages/commit/cb4c843f1feb13347fb472aa71d18f450a0a5413).

https://github.com/sagemath/sage/pull/36109 was merged a few hours after you merged this, and that is the only "long" part of the move to cython 3.

The actual patch needed on top of that to be able to use cython 3 is just 6 lines  (plus 19 lines in doctests due to some changed strings). See #45887 for details.

**Pull request closed without merging**